bundle/bundler/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+        actor_type: actor::builtin::Type,
+        forced_cid: Option<Cid>,
+        bytecode: &[u8],
+    ) -> Result<Cid> {
+        let cid = match forced_cid {
+            Some(cid) => {
+                self.blockstore.put_keyed(&cid, bytecode).with_context(|| {
+                    format!(
+                        "failed to put bytecode for actor {:?} into blockstore",
+                        actor_type
+                    )
+                })?;
+                cid
+            }
+            None => {
+                let blk = &Block {
+                    codec: IPLD_RAW,
+                    data: bytecode,
+                };
+                self.blockstore
+                    .put(Code::Blake2b256, blk)
+                    .with_context(|| {
+                        format!(
+                            "failed to put bytecode for actor {:?} into blockstore",
+                            actor_type
+                        )
+                    })?
+            }
+        };
+        self.added.insert(actor_type, cid);
+        Ok(cid)
+    }
+
+    /// Adds bytecode from a file.
+    pub fn add_from_file<P: AsRef<Path>>(
+        &mut self,
+        actor_type: actor::builtin::Type,
+        forced_cid: Option<Cid>,
+        bytecode_path: P,
+    ) -> Result<Cid> {
+        let bytecode = std::fs::read(bytecode_path).context("failed to open bytecode file")?;
+        self.add_from_bytes(actor_type, forced_cid, bytecode.as_slice())
+    }
+
+    /// Commits the added bytecode entries and writes the CAR file to disk.
+    pub fn finish(self) -> Result<()> {
+        block_on(self.write_car())
+    }
+
+    async fn write_car(self) -> Result<()> {
+        let mut out = async_std::fs::File::create(&self.bundle_dst).await?;
+
+        // Invert the actor index so that it's CID => Type.
+        let manifest: Manifest = self
+            .added
+            .into_iter()
+            .map(|(typ, cid)| (cid, typ))
+            .collect();
+
+        let manifest_bytes = serde_cbor::to_vec(&manifest)?;
+        let root = self.blockstore.put(
+            Code::Blake2b256,
+            &Block {
+                codec: DAG_CBOR,
+                data: &manifest_bytes,
+            },
+        )?;
+
+        // Create a CAR header.
+        let car = CarHeader {
+            roots: vec![root],
+            version: 1,
+        };
+
+        let (tx, mut rx) = bounded(16);
+        let write_task =
+            task::spawn(async move { car.write_stream_async(&mut out, &mut rx).await.unwrap() });
+
+        // Add the root payload.
+        tx.send((root, manifest_bytes)).await.unwrap();
+
+        // Add the bytecodes.
+        for cid in manifest.iter().map(|(cid, _)| cid) {
+            println!("adding cid {} to bundle CAR", cid.to_string());
+            let data = self.blockstore.get(cid).unwrap().unwrap();
+            tx.send((*cid, data)).await.unwrap();
+        }
+
+        drop(tx);
+
+        write_task.await;
```

Comment:
:vomiting_face: this is where go shines.

Reply:
:-( Admittedly there are probably cleaner ways to implement this, but I didn't spend too much time making it nice.